[NativeAOT] Enable Reflection library tests in Linux-ARM64

This test hang in CI Linux-arm64 container without much debugging info and currently its [disabled ](https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj#L395-L396)in our CI runs.

I'm still trying to reproduce this NativeAOT library test hang on Linux, but it's turning out to be very difficult for me. I was able to get reliable test runs (always passing) on Ubuntu 20.04 x64, but after looking at the logs, it looks like the hang is actually reproducing on Ubuntu 18.04 arm64. So I thought I'd try that OS/arch to see if I can actually hit the hang locally.

I saw that there is a docker image `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-20220427172132-97d8652` being used in CI, but I am having a lot of issues with that. Is there a script or build file that I can use to set up the test environment in a container? I've tried to piece it together from the CI logs, but it's not straight forward how everything is getting set up in Helix. 


I also just tried spinning up my own VM and setting it up with these commands, based partially on instructions [here](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md). Is there an issue with my setup steps below?
```
sudo bash -c "$(wget -O - https://apt.kitware.com/kitware-archive.sh)"
sudo bash -c "$(wget -O - https://apt.llvm.org/llvm.sh)"

sudo apt-get install -y cmake llvm-9 clang build-essential python curl git lldb-6.0 liblldb-6.0-dev libunwind8 libunwind8-dev gettext libicu-dev liblttng-ust-dev libssl-dev libnuma-dev libkrb5-dev zlib1g-dev ninja-build git

git clone https://github.com/dotnet/runtime.git
cd runtime

./build.sh clr.alljits+clr.tools+clr.nativeaotlibs+clr.nativeaotruntime+libs+libs.tests -rc Release -lc Release /p:TestNativeAot=true
```

When set up that way, it results in a `Segmentation fault (core dumped)` during build. How can I examine the seg fault deeper to see what the problem is? And why would my local VM be throwing it, but the same OS/Arch on Helix isn't. I probably have some dependency installed wrong, but I'm not sure which one. 

```
109:5>Done Building Project "/home/ubuntu/code/runtime/src/libraries/System.Runtime.Caching/tests/System.Runtime.Caching.Tests.csproj" (default targets).
  65:6>IlcCompile:
         Segmentation fault (core dumped)
  65:6>/home/ubuntu/code/runtime/artifacts/bin/coreclr/Linux.arm64.Release/build/Microsoft.NETCore.Native.targets(270,5): error MSB3073: The command ""/home/ubuntu/code/runtime/artifacts/bin/coreclr/Linux.arm64.Release/ilc/ilc" @"/home/ubuntu/code/runtime/artifacts/obj/System.Linq.Expressions.Tests/Release/net7.0/native/System.Linq.Expressions.Tests.ilc.rsp"" exited with code 139. [/home/ubuntu/code/runtime/src/libraries/System.Linq.Expressions/tests/System.Linq.Expressions.Tests.csproj]
       Done executing task "Exec" -- FAILED.
  65:6>Done building target "IlcCompile" in project "System.Linq.Expressions.Tests.csproj" -- FAILED.
  65:6>Done Building Project "/home/ubuntu/code/runtime/src/libraries/System.Linq.Expressions/tests/System.Linq.Expressions.Tests.csproj" (default targets) -- FAILED.
```

--- a/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs
@@ -118,7 +118,7 @@ namespace System.Threading
 
             /// <summary>
             /// Callers must ensure to clear the array after use. Once <see cref="RegisterWait(int, bool, bool)"/> is called (followed
-            /// by a call to <see cref="Wait(int, bool, bool)"/>, the array will be cleared automatically.
+            /// by a call to <see cref="Wait(int, bool, bool, ref LockHolder)"/>, the array will be cleared automatically.
             /// </summary>
             public WaitableObject?[] GetWaitedObjectArray(int requiredCapacity)
             {
@@ -276,7 +276,7 @@ namespace System.Threading
                 }
             }
 
-            public int Wait(int timeoutMilliseconds, bool interruptible, bool isSleep)
+            public int Wait(int timeoutMilliseconds, bool interruptible, bool isSleep, ref LockHolder lockHolder)
             {
                 if (isSleep)
                 {
@@ -299,7 +299,7 @@ namespace System.Threading
                 _waitMonitor.Acquire();
                 if (!isSleep)
                 {
-                    s_lock.Release();
+                    lockHolder.Dispose();
                 }
 
                 Debug.Assert(_waitedObjectIndexThatSatisfiedWait < 0);
@@ -404,11 +404,12 @@ namespace System.Threading
                     return;
                 }
 
+                LockHolder dummyLockHolder = default;
                 int waitResult =
                     Thread
                         .CurrentThread
                         .WaitInfo
-                        .Wait(timeoutMilliseconds, interruptible, isSleep: true);
+                        .Wait(timeoutMilliseconds, interruptible, isSleep: true, ref dummyLockHolder);
                 Debug.Assert(waitResult == WaitHandle.WaitTimeout);
             }
 

--- a/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs
@@ -125,7 +125,7 @@ namespace System.Threading
         // around and keep track of whether the lock still needs to be released.
         public struct LockHolder
         {
-            LowLevelLock? _lock;
+            private LowLevelLock? _lock;
 
             public LockHolder(LowLevelLock l)
             {

--- a/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs
@@ -120,9 +120,44 @@ namespace System.Threading
     {
         private static readonly LowLevelLock s_lock = new LowLevelLock();
 
+        // Exception handling may use the WaitSubsystem. It means that we need to release the WaitSubsystem
+        // lock before throwing any exceptions to avoid deadlocks. LockHolder allows us to pass the lock state
+        // around and keep track of whether the lock still needs to be released.
+        public struct LockHolder
+        {
+            LowLevelLock? _lock;
+
+            public LockHolder(LowLevelLock l)
+            {
+                l.Acquire();
+                _lock = l;
+            }
+
+            public void Dispose()
+            {
+                if (_lock != null)
+                {
+                    _lock.Release();
+                    _lock = null;
+                }
+            }
+        }
+
         private static SafeWaitHandle NewHandle(WaitableObject waitableObject)
         {
-            IntPtr handle = HandleManager.NewHandle(waitableObject);
+            IntPtr handle = IntPtr.Zero;
+            try
+            {
+                handle = HandleManager.NewHandle(waitableObject);
+            }
+            finally
+            {
+                if (handle == IntPtr.Zero)
+                {
+                    waitableObject.OnDeleteHandle();
+                }
+            }
+
             SafeWaitHandle? safeWaitHandle = null;
             try
             {
@@ -172,8 +207,7 @@ namespace System.Threading
             // between adding the mutex to the named object table and initially acquiring it.
             // To avoid the possibility of another thread retrieving the mutex via its name
             // before we managed to acquire it, we perform both steps while holding s_lock.
-            s_lock.Acquire();
-            bool holdingLock = true;
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
                 WaitableObject? waitableObject = WaitableObject.CreateNamedMutex_Locked(name, out createdNew);
@@ -191,18 +225,13 @@ namespace System.Threading
                 // by the thread. See <see cref="ThreadWaitInfo.LockedMutexesHead"/>. So, acquire the lock only after all
                 // possibilities for exceptions have been exhausted.
                 ThreadWaitInfo waitInfo = Thread.CurrentThread.WaitInfo;
-                int status = waitableObject.Wait_Locked(waitInfo, timeoutMilliseconds: 0, interruptible: false, prioritize: false);
+                int status = waitableObject.Wait_Locked(waitInfo, timeoutMilliseconds: 0, interruptible: false, prioritize: false, ref lockHolder);
                 Debug.Assert(status == 0);
-                // Wait_Locked has already released s_lock, so we no longer hold it here.
-                holdingLock = false;
                 return safeWaitHandle;
             }
             finally
             {
-                if (holdingLock)
-                {
-                    s_lock.Release();
-                }
+                lockHolder.Dispose();
             }
         }
 
@@ -227,14 +256,14 @@ namespace System.Threading
         {
             Debug.Assert(waitableObject != null);
 
-            s_lock.Acquire();
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
-                waitableObject.SignalEvent();
+                waitableObject.SignalEvent(ref lockHolder);
             }
             finally
             {
-                s_lock.Release();
+                lockHolder.Dispose();
             }
         }
 
@@ -247,14 +276,14 @@ namespace System.Threading
         {
             Debug.Assert(waitableObject != null);
 
-            s_lock.Acquire();
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
-                waitableObject.UnsignalEvent();
+                waitableObject.UnsignalEvent(ref lockHolder);
             }
             finally
             {
-                s_lock.Release();
+                lockHolder.Dispose();
             }
         }
 
@@ -269,14 +298,14 @@ namespace System.Threading
             Debug.Assert(waitableObject != null);
             Debug.Assert(count > 0);
 
-            s_lock.Acquire();
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
-                return waitableObject.SignalSemaphore(count);
+                return waitableObject.SignalSemaphore(count, ref lockHolder);
             }
             finally
             {
-                s_lock.Release();
+                lockHolder.Dispose();
             }
         }
 
@@ -289,14 +318,14 @@ namespace System.Threading
         {
             Debug.Assert(waitableObject != null);
 
-            s_lock.Acquire();
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
-                waitableObject.SignalMutex();
+                waitableObject.SignalMutex(ref lockHolder);
             }
             finally
             {
-                s_lock.Release();
+                lockHolder.Dispose();
             }
         }
 
@@ -412,38 +441,30 @@ namespace System.Threading
             Debug.Assert(timeoutMilliseconds >= -1);
 
             ThreadWaitInfo waitInfo = Thread.CurrentThread.WaitInfo;
-            bool waitCalled = false;
-            s_lock.Acquire();
+            LockHolder lockHolder = new LockHolder(s_lock);
             try
             {
                 // A pending interrupt does not signal the specified handle
                 if (interruptible && waitInfo.CheckAndResetPendingInterrupt)
                 {
+                    lockHolder.Dispose();
                     throw new ThreadInterruptedException();
                 }
 
                 try
                 {
-                    waitableObjectToSignal.Signal(1);
+                    waitableObjectToSignal.Signal(1, ref lockHolder);
                 }
                 catch (SemaphoreFullException ex)
                 {
+                    s_lock.VerifyIsNotLocked();
                     throw new InvalidOperationException(SR.Threading_WaitHandleTooManyPosts, ex);
                 }
-                waitCalled = true;
-                return waitableObjectToWaitOn.Wait_Locked(waitInfo, timeoutMilliseconds, interruptible, prioritize);
+                return waitableObjectToWaitOn.Wait_Locked(waitInfo, timeoutMilliseconds, interruptible, prioritize, ref lockHolder);
             }
             finally
             {
-                // Once the wait function is called, it will release the lock
-                if (waitCalled)
-                {
-                    s_lock.VerifyIsNotLocked();
-                }
-                else
-                {
-                    s_lock.Release();
-                }
+                lockHolder.Dispose();
             }
         }
 

--- a/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs
@@ -308,22 +308,28 @@ namespace System.Threading
 
                 Debug.Assert(timeoutMilliseconds >= -1);
 
-                s_lock.Acquire();
-
-                if (interruptible && waitInfo.CheckAndResetPendingInterrupt)
+                var lockHolder = new LockHolder(s_lock);
+                try
                 {
-                    s_lock.Release();
-                    throw new ThreadInterruptedException();
-                }
+                    if (interruptible && waitInfo.CheckAndResetPendingInterrupt)
+                    {
+                        lockHolder.Dispose();
+                        throw new ThreadInterruptedException();
+                    }
 
-                return Wait_Locked(waitInfo, timeoutMilliseconds, interruptible, prioritize);
+                    return Wait_Locked(waitInfo, timeoutMilliseconds, interruptible, prioritize, ref lockHolder);
+                }
+                finally
+                {
+                    lockHolder.Dispose();
+                }
             }
 
             /// <summary>
             /// This function does not check for a pending thread interrupt. Callers are expected to do that soon after
             /// acquiring <see cref="s_lock"/>.
             /// </summary>
-            public int Wait_Locked(ThreadWaitInfo waitInfo, int timeoutMilliseconds, bool interruptible, bool prioritize)
+            public int Wait_Locked(ThreadWaitInfo waitInfo, int timeoutMilliseconds, bool interruptible, bool prioritize, ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
                 Debug.Assert(waitInfo != null);
@@ -332,50 +338,39 @@ namespace System.Threading
                 Debug.Assert(timeoutMilliseconds >= -1);
                 Debug.Assert(!interruptible || !waitInfo.CheckAndResetPendingInterrupt);
 
-                bool needToWait = false;
-                try
+                if (IsSignaled)
                 {
-                    if (IsSignaled)
-                    {
-                        bool isAbandoned = IsAbandonedMutex;
-                        AcceptSignal(waitInfo);
-                        return isAbandoned ? WaitHandle.WaitAbandoned : WaitHandle.WaitSuccess;
-                    }
-
-                    if (IsMutex && _ownershipInfo != null && _ownershipInfo.Thread == waitInfo.Thread)
-                    {
-                        if (!_ownershipInfo.CanIncrementReacquireCount)
-                        {
-                            throw new OverflowException(SR.Overflow_MutexReacquireCount);
-                        }
-                        _ownershipInfo.IncrementReacquireCount();
-                        return WaitHandle.WaitSuccess;
-                    }
-
-                    if (timeoutMilliseconds == 0)
-                    {
-                        return WaitHandle.WaitTimeout;
-                    }
-
-                    WaitableObject?[] waitableObjects = waitInfo.GetWaitedObjectArray(1);
-                    waitableObjects[0] = this;
-                    waitInfo.RegisterWait(1, prioritize, isWaitForAll: false);
-                    needToWait = true;
+                    bool isAbandoned = IsAbandonedMutex;
+                    AcceptSignal(waitInfo);
+                    return isAbandoned ? WaitHandle.WaitAbandoned : WaitHandle.WaitSuccess;
                 }
-                finally
+
+                if (IsMutex && _ownershipInfo != null && _ownershipInfo.Thread == waitInfo.Thread)
                 {
-                    // Once the wait function is called, it will release the lock
-                    if (!needToWait)
+                    if (!_ownershipInfo.CanIncrementReacquireCount)
                     {
-                        s_lock.Release();
+                        lockHolder.Dispose();
+                        throw new OverflowException(SR.Overflow_MutexReacquireCount);
                     }
+                    _ownershipInfo.IncrementReacquireCount();
+                    return WaitHandle.WaitSuccess;
                 }
+
+                if (timeoutMilliseconds == 0)
+                {
+                    return WaitHandle.WaitTimeout;
+                }
+
+                WaitableObject?[] waitableObjects = waitInfo.GetWaitedObjectArray(1);
+                waitableObjects[0] = this;
+                waitInfo.RegisterWait(1, prioritize, isWaitForAll: false);
 
                 return
                     waitInfo.Wait(
                         timeoutMilliseconds,
                         interruptible,
-                        isSleep: false);
+                        isSleep: false,
+                        ref lockHolder);
             }
 
             public static int Wait(
@@ -396,12 +391,12 @@ namespace System.Threading
                 Debug.Assert(count > 1);
                 Debug.Assert(timeoutMilliseconds >= -1);
 
-                bool needToWait = false;
-                s_lock.Acquire();
+                var lockHolder = new LockHolder(s_lock);
                 try
                 {
                     if (interruptible && waitInfo.CheckAndResetPendingInterrupt)
                     {
+                        lockHolder.Dispose();
                         throw new ThreadInterruptedException();
                     }
 
@@ -431,6 +426,7 @@ namespace System.Threading
                                 {
                                     if (!ownershipInfo.CanIncrementReacquireCount)
                                     {
+                                        lockHolder.Dispose();
                                         throw new OverflowException(SR.Overflow_MutexReacquireCount);
                                     }
                                     ownershipInfo.IncrementReacquireCount();
@@ -465,6 +461,7 @@ namespace System.Threading
                                 {
                                     if (!ownershipInfo.CanIncrementReacquireCount)
                                     {
+                                        lockHolder.Dispose();
                                         throw new OverflowException(SR.Overflow_MutexReacquireCount);
                                     }
                                     continue;
@@ -494,6 +491,7 @@ namespace System.Threading
 
                             if (isAnyAbandonedMutex)
                             {
+                                lockHolder.Dispose();
                                 throw new AbandonedMutexException();
                             }
                             return WaitHandle.WaitSuccess;
@@ -506,11 +504,14 @@ namespace System.Threading
                     }
 
                     waitableObjects = null; // no need to clear this anymore, RegisterWait / Wait will take over from here
+
                     waitInfo.RegisterWait(count, prioritize, waitForAll);
-                    needToWait = true;
+                    return waitInfo.Wait(timeoutMilliseconds, interruptible, isSleep: false, ref lockHolder);
                 }
                 finally
                 {
+                    lockHolder.Dispose();
+
                     if (waitableObjects != null)
                     {
                         for (int i = 0; i < count; ++i)
@@ -518,15 +519,7 @@ namespace System.Threading
                             waitableObjects[i] = null;
                         }
                     }
-
-                    // Once the wait function is called, it will release the lock
-                    if (!needToWait)
-                    {
-                        s_lock.Release();
-                    }
                 }
-
-                return waitInfo.Wait(timeoutMilliseconds, interruptible, isSleep: false);
             }
 
             public static bool WouldWaitForAllBeSatisfiedOrAborted(
@@ -623,7 +616,7 @@ namespace System.Threading
                 }
             }
 
-            public void Signal(int count)
+            public void Signal(int count, ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
                 Debug.Assert(count > 0);
@@ -641,17 +634,17 @@ namespace System.Threading
                         break;
 
                     case WaitableObjectType.Semaphore:
-                        SignalSemaphore(count);
+                        SignalSemaphore(count, ref lockHolder);
                         break;
 
                     default:
                         Debug.Assert(count == 1);
-                        SignalMutex();
+                        SignalMutex(ref lockHolder);
                         break;
                 }
             }
 
-            public void SignalEvent()
+            public void SignalEvent(ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
 
@@ -666,6 +659,7 @@ namespace System.Threading
                         break;
 
                     default:
+                        lockHolder.Dispose();
                         WaitHandle.ThrowInvalidHandleException();
                         break;
                 }
@@ -721,12 +715,13 @@ namespace System.Threading
                 _signalCount = 1;
             }
 
-            public void UnsignalEvent()
+            public void UnsignalEvent(ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
 
                 if (!IsEvent)
                 {
+                    lockHolder.Dispose();
                     WaitHandle.ThrowInvalidHandleException();
                 }
 
@@ -736,13 +731,14 @@ namespace System.Threading
                 }
             }
 
-            public int SignalSemaphore(int count)
+            public int SignalSemaphore(int count, ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
                 Debug.Assert(count > 0);
 
                 if (!IsSemaphore)
                 {
+                    lockHolder.Dispose();
                     WaitHandle.ThrowInvalidHandleException();
                 }
 
@@ -750,6 +746,7 @@ namespace System.Threading
                 Debug.Assert(oldSignalCount <= _maximumSignalCount);
                 if (count > _maximumSignalCount - oldSignalCount)
                 {
+                    lockHolder.Dispose();
                     throw new SemaphoreFullException();
                 }
 
@@ -776,17 +773,19 @@ namespace System.Threading
                 return oldSignalCount;
             }
 
-            public void SignalMutex()
+            public void SignalMutex(ref LockHolder lockHolder)
             {
                 s_lock.VerifyIsLocked();
 
                 if (!IsMutex)
                 {
+                    lockHolder.Dispose();
                     WaitHandle.ThrowInvalidHandleException();
                 }
 
                 if (IsSignaled || _ownershipInfo!.Thread != Thread.CurrentThread)
                 {
+                    lockHolder.Dispose();
                     throw new ApplicationException(SR.Arg_SynchronizationLockException);
                 }
 

